.github/workflows/ci.yaml

```diff
@@ -57,6 +57,75 @@ jobs:
         run: nox -e ci_checks_max pytest_min
         timeout-minutes: 10
 
+  nox-cross-arch:
+    name: Cross-arch tests with nox
+    needs: ["nox"]
+    if: github.event_name == 'merge_group'
+    strategy:
+      fail-fast: false
+      # Before adding new items to this matrix, make sure that a dockerfile
+      # exists for the combination of items in the matrix.
+      # Refer to .github/containers/nox-cross-arch/README.md to learn how to
+      # add and name new dockerfiles.
+      matrix:
+        arch:
+          - arm64
+        os:
+          - ubuntu-20.04
+        python:
+          - "3.11"
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: Fetch sources
+        uses: actions/checkout@v3
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      # This is a workaround to prevent the cache from growing indefinitely.
+      # https://docs.docker.com/build/ci/github-actions/cache/#local-cache
+      # https://github.com/docker/build-push-action/issues/252
+      # https://github.com/moby/buildkit/issues/1896
+      - name: Cache container layers
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-nox-${{ matrix.arch }}-${{ matrix.os }}-${{ matrix.python }}
+
+      - name: Build image
+        uses: docker/build-push-action@v4
+        with:
+          context: .github/containers/nox-cross-arch
+          file: .github/containers/nox-cross-arch/${{ matrix.arch }}-${{ matrix.os }}-python-${{ matrix.python }}.Dockerfile
+          platforms: linux/${{ matrix.arch }}
+          tags: localhost/nox-cross-arch:latest
+          push: false
+          load: true
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+
+      # Refer to the workaround mentioned above
+      - name: Move cache
+        run: |
+          rm -rf /tmp/.buildx-cache
+          mv /tmp/.buildx-cache-new /tmp/.buildx-cache
+
+      - name: Run nox
+        run: |
+          docker run \
+            --rm \
+            -v $(pwd):/${{ github.workspace }} \
+            -w ${{ github.workspace }} \
+            --net=host \
+            --platform linux/${{ matrix.arch }} \
+            localhost/nox-cross-arch:latest \
+            nox -e pytest_min
+        timeout-minutes: 30
```

Comment:
I would also split these steps to match what we now have in the `nox` job, so we get the output of `pip freeze` to know exactly which deps are being installed for debugging purposes.

Reply:
> I would also split these steps to match what we now have in the nox job

So, run the other step before the docker run command? What happens if some of the dependencies are architecture dependent? Is that even possible for the SDK? Should the tests fail then? (that could be a legit requirement)

Reply:
Ah, right. This is more complicated because you are running `docker` manually. This might work if we create a virtualenv in the current directory (so the `pip install` shouldn't be done in the `Dockerfile` entrypoint, it should be done in the GH workflow), then both venvs, then one we created and the ones created by nox will be mounted as a volume and preserved across docker runs.

It would be ideal to just be able to [run the whole job in the container directly](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container), but I guess we can't because of the QEMU part. Is QEMU needed for building the image, running it or both?

Another way that might be easier is to run a step with `docker run -d ...`, add a final step with `if: always()` that will kill the container we just ran as a cleanup step (to be nice), and then to split the intermediate steps use `docker exec` on the daemonized container running in the background.

Anyway, I thought this was going to be much trivial, I didn't have in mind the explicit docker part, I'm not sure if any of the above work is worth it just to get the extra debug info and a some speedup, I guess it depends on how much keen you are in trying one of the approaches :)

Reply:
> Is QEMU needed for building the image, running it or both?

QEMU is required for running the image. We need to emulate an arm64 platform here, and that's what QEMU does for us.

Reply:
So, after a few (lengthy) tests and trials, it turns out that caching dependencies from pyproject.toml is non-trivial. It is not possible to install just the dependencies [yet](https://github.com/pypa/pip/issues/11440). So, just installing the dependencies in the image doesn't work that way. You could `COPY` the whole project into the image and run it, but then the layer gets rebuilt every time, defeating the purpose.

The next thing that could be cached is the `.nox` directory, which is trivial, and has been added now.

Once the docker image and the .nox directory is in cache, the job takes about 13-15 minutes to run. Although a lot of time is taken by the last tests (`src/frequenz/.../*.py`).

Reply:
>Another way that might be easier is to run a step with docker run -d ..., add a final step with if: always() that will kill the container we just ran as a cleanup step (to be nice), and then to split the intermediate steps use docker exec on the daemonized container running in the background.

While running commands in the docker container technically works, it does not help with caching cross-arch packages. Maybe I misunderstood something here?

Reply:
> Yeah, I wasn't talking about caching at all, it was just a small thing about having a separate step to be able to know which dependencies were used exactly when running the tests, to be able to debug weird cases where stuff work locally but not in the CI or viceversa.

Oh, right, and this doesn't need to be checked more than once at all. There is too much mess to be cleaned here :(

As a workaround, if this saves a lot of time, you could try running `nox -e pytest_xxx -- tests`, this should not run the testing of examples in `src/`.

Reply:
> While running commands in the docker container technically works, it does not help with caching cross-arch packages. Maybe I misunderstood something here?

The main point about that was not caching actually, but just printing the exact version of installed deps, for debugging purposes only, but I just realized that we can do that in only once step too, it is just a bit more annoying to go look for the output. I put the suggestion in the review.

Reply:
okay, I'll add `pip freeze` to the commands in the container.

Reply:
Done